### PR TITLE
Jenkins api integration

### LINKED ISSUE #4: Import Jenkins data to Firestore
Develop a program that uses the Jenkins API to import all builds for a given job to Firebase. This should be idempotent. The program will run on a schedule and constantly update Firestore with the latest builds.

Name | Description
| --- | --- |
buildId | unique identifier of the build
buildDate | date the build was started
buildQueued | the amount of time the build spent waiting to execute
buildDuration | the amount of time the build took to execute (does not include queue time)
buildSuccessful | boolean for if the build was successful or not

### PR DESCRIPTION
**Pull request by vova-beloded**
## Checklist

- [x] Documented
- [x] Unit tested

### CODE REVIEW

File: metrics_ci_integration/lib/jenkins/model/jenkins_query_limits.dart [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+  factory JenkinsQueryLimits.fromQuery(String query) {
+    if (query.startsWith('{') && query.endsWith('}')) {
+      final limits = query
+          .split(',')
+          .map((substring) => substring.replaceAll(RegExp(r'[\{\}]'), ''))
+          .toList();
+
+      final limitsValid = (limits.length == 1 &&
+              _checkValueString(limits.first, allowEmpty: false)) ||
+          (limits.length < 3 && limits.every(_checkValueString));
```
Comment (vova-beloded) [inline metrics_ci_integration/lib/jenkins/model/jenkins_query_limits.dart]:
```suggestion
            final limitsValid = limits.length < 3 &&
          limits.every((limit) =>
              _checkValueString(limit, allowEmpty: limits.length != 1));
```
In the current implementation, the limits are valid if the query equals to '{}'. Maybe we can fix it like this and simplify this statement?

File: metrics_ci_integration/lib/jenkins/client/jenkins_client.dart [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+  static const String buildTreeQuery =
+      'number,duration,timestamp,result,url,artifacts[$artifactsTreeQuery]';
+
+  /// The part of the `tree` query parameter that stands for [JenkinsJob]'s
+  /// properties to fetch.
+  @visibleForTesting
+  static const String jobBaseTreeQuery = 'name,fullName,url';
+
+  /// The part of the `tree` query parameter that extends common [JenkinsJob]'s
+  /// properties to fetch with `jobs` and `builds` to detect a job type.
```
Comment (vova-beloded) [inline metrics_ci_integration/lib/jenkins/client/jenkins_client.dart]:
```suggestion
   /// The part of the `tree` query parameter that extends common [JenkinsJob]'s 
  /// properties to fetch the job with `sub-jobs` and `builds`
  /// to be able to detect a job type.
```

File: metrics_ci_integration/lib/jenkins/model/jenkins_build.dart [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    this.artifacts,
+  });
+
+  /// Creates an instance of Jenkins build from the decoded JSON object.
+  ///
+  /// Returns `null` if [json] is `null`.
+  factory JenkinsBuild.fromJson(Map<String, dynamic> json) {
+    if (json == null) return null;
+
+    final timestamp = json['timestamp'] == null
```
Comment (vova-beloded) [inline metrics_ci_integration/lib/jenkins/model/jenkins_build.dart]: Maybe we can replace this with int.tryParse() to make it more reliable?

File: metrics_ci_integration/lib/jenkins/model/jenkins_build.dart [hunk 4]
```diff
@@ -0,0 +1,95 @@
+import 'package:ci_integration/jenkins/model/jenkins_build_artifact.dart';
+import 'package:ci_integration/jenkins/model/jenkins_building_job.dart';
+import 'package:equatable/equatable.dart';
+
+/// A class representing a single Jenkins build.
+class JenkinsBuild extends Equatable {
```
Comment (vova-beloded) [inline metrics_ci_integration/lib/jenkins/model/jenkins_build.dart]: Maybe we can move the 'model' package into the 'client' package? Will be happy to discuss it with you.

File: metrics_ci_integration/lib/jenkins/model/jenkins_building_job.dart [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      ),
+      builds: JenkinsBuild.listFromJson(json['builds'] as List<dynamic>),
+    );
+  }
+
+  /// Converts object into the [Map].
+  ///
+  /// The resulting map will include only non-null fields of an object it
+  /// represents and can be encoded to a JSON object.
+  /// Populates [JenkinsJob.toJson] with building job specific fields.
```
Comment (illia-romanenko) [inline metrics_ci_integration/lib/jenkins/model/jenkins_building_job.dart]:
```suggestion

```

File: metrics_ci_integration/lib/jenkins/model/jenkins_multi_branch_job.dart [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    return JenkinsMultiBranchJob(
+      name: json['name'] as String,
+      fullName: json['fullName'] as String,
+      url: json['url'] as String,
+      jobs: JenkinsJob.listFromJson(json['jobs'] as List<dynamic>),
+    );
+  }
+
+  /// Converts object into the [Map].
+  /// The result can be encoded to a JSON object.
```
Comment (illia-romanenko) [inline metrics_ci_integration/lib/jenkins/model/jenkins_multi_branch_job.dart]: Should we revisit it to toMap, fromMap?

File: metrics_ci_integration/lib/jenkins/model/jenkins_multi_branch_job.dart [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  ///
+  /// The resulting map will include only non-null fields of an object it
+  /// represents and can be encoded to a JSON object.
+  /// Populates [JenkinsJob.toJson] with [jobs].
```
Comment (illia-romanenko) [inline metrics_ci_integration/lib/jenkins/model/jenkins_multi_branch_job.dart]:
```suggestion
```

File: metrics_ci_integration/lib/jenkins/model/jenkins_result.dart [hunk 8]
```diff
@@ -0,0 +1,39 @@
+/// A class containing a result for Jenkins API interaction.
+class JenkinsResult<T> {
+  /// Used to indicate that interaction is failed.
+  final bool _isError;
+
+  /// Contains message with a result of interaction.
+  /// Contains error message if interaction failed.
```
Comment (illia-romanenko) [inline metrics_ci_integration/lib/jenkins/model/jenkins_result.dart]:
```suggestion
```

File: metrics_ci_integration/pubspec.yaml [hunk 9]
```diff
@@ -0,0 +1,17 @@
+name: ci_integration
+description: A sample command-line application.
+# version: 1.0.0
+# homepage: https://www.example.com
```
Comment (illia-romanenko) [inline metrics_ci_integration/pubspec.yaml]: Could you please a bit more details here, description that helps to understand this tool?

File: metrics_ci_integration/test/test_utils/api_mock_server/request_handler.dart [hunk 10]
```diff
@@ -0,0 +1,100 @@
+import 'dart:io';
+
+/// Handle [HttpRequest] callback definition
+typedef RequestHandleDispatcher = Future<void> Function(HttpRequest);
```
Comment (illia-romanenko) [inline metrics_ci_integration/test/test_utils/api_mock_server/request_handler.dart]: We should try to reuse api_mock_server from guardian here. We can follow the same approach that we took for yaml - moving api_mock_server to the top level for reuse.

File: metrics_ci_integration/test/jenkins/model/jenkins_result_test.dart [hunk 11]
```diff
@@ -0,0 +1,25 @@
+import 'package:ci_integration/jenkins/model/jenkins_result.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('JenkinsResult', () {
+    const failed = JenkinsResult.error(message: 'Error');
+    const success = JenkinsResult.success(message: 'Hooray!');
```
Comment (illia-romanenko) [inline metrics_ci_integration/test/jenkins/model/jenkins_result_test.dart]: Should we also add tests for `result`?

File: metrics_ci_integration/test/common/authorization/authorization_base_test.dart [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        final authorization = AuthorizationBaseTestbed(header, token);
+        final expected = {header: token};
+        final actual = authorization.toMap();
+
+        expect(actual, equals(expected));
+      },
+    );
+
+    test(
+      'should throw ArgumentError creating a new instance with null header',
```
Comment (illia-romanenko) [inline metrics_ci_integration/test/common/authorization/authorization_base_test.dart]: @vova-beloded @vladlaba  I think we need to decide about test description style. Should we make it a regular sentence or it should be informal like here:
```suggestion
      'should throw ArgumentError creating a new instance with null header',
```

or 
```suggestion
      'Should throw ArgumentError creating a new instance with null header.',
```
I see it's different now across our monorepo so it's probably better to choose one option.

File: metrics_ci_integration/lib/jenkins/model/jenkins_job.dart [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+  final String url;
+
+  @override
+  List<Object> get props => [name, fullName, url];
+
+  const JenkinsJob({
+    @required this.name,
+    this.fullName,
+    this.url,
+  });
```
Comment (illia-romanenko) [inline metrics_ci_integration/lib/jenkins/model/jenkins_job.dart]: Should we assert that name is not null?

File: metrics_ci_integration/lib/jenkins/model/jenkins_build.dart (hunk 3 above)
Comment (vladlaba) [reply metrics_ci_integration/lib/jenkins/model/jenkins_build.dart]: I'm not sure I understand what do you mean. 

`int.tryParse()` throws if argument is `null`. Do you mean to change a `timestamp` property type to `int` instead of `DateTime` to just use `json['timestamp'] as int`?

File: metrics_ci_integration/lib/jenkins/model/jenkins_build_artifact.dart [hunk 14]
```diff
@@ -0,0 +1,57 @@
+import 'package:ci_integration/jenkins/model/jenkins_build.dart';
+import 'package:equatable/equatable.dart';
+
+/// A class representing a single artifact of a [JenkinsBuild].
+class JenkinsBuildArtifact extends Equatable {
+  /// A name of this artifact file.
+  final String fileName;
+
+  /// A path to this artifact.
+  final String relativePath;
```
[the remark this answers is not in the document]
Comment (vladlaba) [reply metrics_ci_integration/lib/jenkins/model/jenkins_build_artifact.dart]: This is a property name from Jenkins API. Since it has no documentation I cannot answer this question. 

I suppose that it stands for the relative path in artifacts tree for the build. From the root to the artifact file.

File: metrics_ci_integration/lib/jenkins/model/jenkins_building_job.dart [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+import 'package:ci_integration/jenkins/model/jenkins_job.dart';
+import 'package:meta/meta.dart';
+
+/// A class representing a Jenkins building job.
+class JenkinsBuildingJob extends JenkinsJob {
+  /// The first build performed within this job.
+  final JenkinsBuild firstBuild;
+
+  /// The current last build performed within this job.
+  final JenkinsBuild lastBuild;
```
[the remark this answers is not in the document]
Comment (vladlaba) [reply metrics_ci_integration/lib/jenkins/model/jenkins_building_job.dart]: In the case when you want to paginate API responses you need to know when it's time to stop - both lastBuild and firstBuild can be useful for that. Moreover, to detect whether there are new builds since the last synchronization the lastBuild property is useful.

File: metrics_ci_integration/lib/jenkins/model/jenkins_job.dart (hunk 13 above)
Comment (vladlaba) [reply metrics_ci_integration/lib/jenkins/model/jenkins_job.dart]: Not sure we should require this at all since the `fullName` property sometimes more useful and we use a `url` property more often.

File: metrics_ci_integration/test/jenkins/model/jenkins_result_test.dart (hunk 11 above)
Comment (vladlaba) [reply metrics_ci_integration/test/jenkins/model/jenkins_result_test.dart]: Not sure we have to test a property. If we try to cover it with tests we'll either check that Dart's generic works properly or `not-null if not-null provided`.

File: metrics_ci_integration/CHANGELOG.md [hunk 16]
```diff
@@ -0,0 +1,3 @@
+## 1.0.0
+
+- Initial version, created by Stagehand
```
[the remark this answers is not in the document]
Comment (vladlaba) [reply metrics_ci_integration/CHANGELOG.md]: https://pub.dev/packages/stagehand 

This file was created automatically.

File: metrics_ci_integration/lib/jenkins/client/jenkins_client.dart [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+    JenkinsQueryLimits limits = const JenkinsQueryLimits.empty(),
+  }) {
+    final url = buildJenkinsApiUrl(
+      multiBranchJob.url,
+      treeQuery: '$jobBaseTreeQuery,jobs[$jobTreeQuery]${limits.toQuery()}',
+    );
+
+    return _handleResponse<JenkinsMultiBranchJob>(
+      _client.get(url, headers: headers),
+      (Map<String, dynamic> json) => JenkinsResult.success(
```
[the remark this answers is not in the document]
Comment (vladlaba) [reply metrics_ci_integration/lib/jenkins/client/jenkins_client.dart]: This is handled by the try-catch block within the `_handleResponse` method - the JenkinsResult.error is returned in case of any error.

File: metrics_ci_integration/lib/jenkins/model/jenkins_build.dart (hunk 3 above)
Comment (vladlaba) [reply metrics_ci_integration/lib/jenkins/model/jenkins_build.dart]: @vova-beloded can you please take a look and clarify?

Comment (vova-beloded) [reply metrics_ci_integration/lib/jenkins/model/jenkins_build.dart]: I mean replace this check with something like this `int.tryParse('${json['timestamp']}') == null` to be sure that json['timestamp'] is a valid integer.

Comment (vladlaba) [reply metrics_ci_integration/lib/jenkins/model/jenkins_build.dart]: I think it is more Jenkins-side to ensure that `timestamp` is an integer.

Comment (vova-beloded) [reply metrics_ci_integration/lib/jenkins/model/jenkins_build.dart]: Ok, then we can leave it as is, I think.

File: metrics_ci_integration/CHANGELOG.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (vladlaba) [reply metrics_ci_integration/CHANGELOG.md]: Wouldn't it be better to delete this file then? What do you think?

[the remark this answers is not in the document]
Comment (illia-romanenko) [reply metrics_ci_integration/CHANGELOG.md]: I think we can delete it for now.

File: metrics_ci_integration/lib/jenkins/client/model/jenkins_build_artifact.dart [hunk 18]
```diff
@@ -0,0 +1,52 @@
+import 'package:ci_integration/jenkins/client/model/jenkins_build.dart';
+import 'package:equatable/equatable.dart';
+
+/// A class representing a single artifact of a [JenkinsBuild].
+class JenkinsBuildArtifact extends Equatable {
+  /// A name of this artifact file.
+  final String fileName;
+
+  /// A path to this artifact.
```
Comment (vova-beloded) [inline metrics_ci_integration/lib/jenkins/client/model/jenkins_build_artifact.dart]: Maybe we should mention that this path points to the artifact that on the Jenkins instance?

File: metrics_ci_integration/lib/jenkins/client/model/jenkins_build_artifact.dart [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+  factory JenkinsBuildArtifact.fromJson(Map<String, dynamic> json) {
+    if (json == null) return null;
+
+    return JenkinsBuildArtifact(
+      fileName: json['fileName'] as String,
+      relativePath: json['relativePath'] as String,
+    );
+  }
+
+  /// Creates a list of artifacts from the [list] of decoded JSON objects.
```
Comment (vova-beloded) [inline metrics_ci_integration/lib/jenkins/client/model/jenkins_build_artifact.dart]: Do we need to mention that such functions return `null` if `list` is null?

File: metrics_ci_integration/lib/jenkins/client/model/jenkins_building_job.dart [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+  const JenkinsBuildingJob({
+    String name,
+    String fullName,
+    String url,
+    this.firstBuild,
+    this.lastBuild,
+    this.builds,
+  }) : super(name: name, fullName: fullName, url: url);
+
+  /// Creates an instance of building job from the decoded JSON object.
```
Comment (vova-beloded) [inline metrics_ci_integration/lib/jenkins/client/model/jenkins_building_job.dart]:
```suggestion
  /// Creates an instance of a building job from the decoded JSON object.
```

File: metrics_ci_integration/lib/jenkins/client/model/jenkins_multi_branch_job.dart [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+
+    return JenkinsMultiBranchJob(
+      name: json['name'] as String,
+      fullName: json['fullName'] as String,
+      url: json['url'] as String,
+      jobs: JenkinsJob.listFromJson(json['jobs'] as List<dynamic>),
+    );
+  }
+
+  /// Converts object into the JSON encodable [Map].
```
Comment (vova-beloded) [inline metrics_ci_integration/lib/jenkins/client/model/jenkins_multi_branch_job.dart]: It seems like we should not add documentation for overridden methods.

File: metrics_ci_integration/lib/jenkins/client/model/jenkins_query_limits.dart [hunk 22]
```diff
@@ -0,0 +1,183 @@
+import 'package:ci_integration/jenkins/client/util/number_validator.dart';
+
+/// A class representing a range specifier for array-type properties in
+/// Jenkins API requests.
+class JenkinsQueryLimits {
+  /// The lower bound of a range specified.
+  final int lower;
+
+  /// The lower bound of a range specified.
```
Comment (vova-beloded) [inline metrics_ci_integration/lib/jenkins/client/model/jenkins_query_limits.dart]:
```suggestion
  /// The upper bound of a range specified.
```

File: metrics_ci_integration/lib/jenkins/client/model/jenkins_result.dart [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+  final String message;
+
+  /// Contains a result of interaction.
+  ///
+  /// Generally, the parsed body from response.
+  final T result;
+
+  bool get isError => _isError;
+
+  bool get isSuccess => !_isError;
```
Comment (vova-beloded) [inline metrics_ci_integration/lib/jenkins/client/model/jenkins_result.dart]: Should we add documentation for these members?

File: metrics_ci_integration/lib/jenkins/client/jenkins_client.dart [hunk 24]
```diff
@@ -0,0 +1,278 @@
...
+import 'package:ci_integration/jenkins/client/model/jenkins_job.dart';
+import 'package:ci_integration/jenkins/client/model/jenkins_multi_branch_job.dart';
+import 'package:ci_integration/jenkins/client/model/jenkins_query_limits.dart';
+import 'package:ci_integration/jenkins/client/model/jenkins_result.dart';
+import 'package:ci_integration/jenkins/client/util/url_utils.dart';
+import 'package:http/http.dart';
+import 'package:meta/meta.dart';
+
+/// A callback for parsing Jenkins API response body.
+typedef BodyParserCallback<T> = JenkinsResult<T> Function(Map<String, dynamic>);
```
Comment (vova-beloded) [inline metrics_ci_integration/lib/jenkins/client/jenkins_client.dart]:
```suggestion
typedef BodyParserCallback<T> = JenkinsResult<T> Function(Map<String, dynamic> body);
```
Can we add the parameter name to be able to generate valid suggestions in IDE? 
I mean that if you'll use suggestion and generate `_handleResponse` call you'll obtain such line of code: 
`_handleResponse(responseFuture, () => null)` which is not valid because closure should take a param.

File: metrics_ci_integration/test/jenkins/client/model/jenkins_build_test.dart [hunk 25]
```diff
@@ -0,0 +1,73 @@
+import 'package:ci_integration/jenkins/client/model/jenkins_build.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group("JenkinsBuild", () {
+    final currentMillisecondsSinceEpoch = DateTime.now().millisecondsSinceEpoch;
```
Comment (vova-beloded) [inline metrics_ci_integration/test/jenkins/client/model/jenkins_build_test.dart]:
```suggestion
    final currentTimestamp = DateTime.now();
```
Can we replace this by DateTime timestamp to make the code a bit more readable?

File: metrics_ci_integration/lib/jenkins/client/model/jenkins_query_limits.dart [hunk 26]
```diff
@@ -0,0 +1,183 @@
...
+  /// right-edged (inclusive) range.
+  ///
+  /// Specifies the `{begin,end+1}` range.
+  /// For example:
+  ///   * `JenkinsQueryLimits.betweenInclusive(3, 10)` stands for the range from
+  ///     the third element (inclusive) to the tenth element (inclusive).
+  factory JenkinsQueryLimits.betweenInclusive(int begin, int end) {
+    NumberValidator.checkRange(begin, end);
+    return JenkinsQueryLimits._(lower: begin, upper: end + 1);
+  }
```
Comment (illia-romanenko) [inline metrics_ci_integration/lib/jenkins/client/model/jenkins_query_limits.dart]: Do we really need such "Inclusive" methods? I think we can always use simple "between" and adjust parameters to get expected results. Maybe we should just remove them?

File: metrics_ci_integration/lib/jenkins/client/model/jenkins_query_limits.dart [hunk 27]
```diff
@@ -0,0 +1,183 @@
...
+    }
+  }
+
+  /// Creates a range specifier for the n-th element
+  /// where `n` is a given [number].
+  ///
+  /// Specifies the `{number}` range.
+  /// For example:
+  ///   * `JenkinsQueryLimits.at(3)` stands for the third element.
+  factory JenkinsQueryLimits.at(int number) {
```
Comment (illia-romanenko) [inline metrics_ci_integration/lib/jenkins/client/model/jenkins_query_limits.dart]: Should we describe contract more clearly here? Positive number?

File: metrics_ci_integration/test/jenkins/client/model/jenkins_result_test.dart [hunk 28]
```diff
@@ -0,0 +1,25 @@
+import 'package:ci_integration/jenkins/client/model/jenkins_result.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group("JenkinsResult", () {
+    const failed = JenkinsResult.error(message: 'Error');
+    const success = JenkinsResult.success(message: 'Hooray!');
```
Comment (illia-romanenko) [inline metrics_ci_integration/test/jenkins/client/model/jenkins_result_test.dart]: Should we also test `result` here?

File: metrics_ci_integration/test/jenkins/client/util/number_validator_test.dart [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+    );
+
+    test(
+      ".checkRange() should throw ArgumentError if the end value is less than the begin one",
+      () {
+        expect(() => NumberValidator.checkRange(2, 1), throwsArgumentError);
+      },
+    );
+
+    test(".checkRange() should validate the given rande", () {
```
Comment (illia-romanenko) [inline metrics_ci_integration/test/jenkins/client/util/number_validator_test.dart]:
```suggestion
    test(".checkRange() should validate the given range", () {
```

File: metrics_ci_integration/test/jenkins/client/model/jenkins_query_limits_test.dart [hunk 30]
```diff
@@ -0,0 +1,159 @@
...
+    test(".empty() should create an empty range specifier", () {
+      const limits = JenkinsQueryLimits.empty();
+      const expected = '';
+      final actual = limits.toQuery();
+
+      expect(actual, equals(expected));
+    });
+
+    test(".at(n) should create a range specifier for n-th element - {n}", () {
+      final limits = JenkinsQueryLimits.at(1);
```
Comment (illia-romanenko) [inline metrics_ci_integration/test/jenkins/client/model/jenkins_query_limits_test.dart]: We should add more tests according to the correct contract (should be positive). This should be applicable to all methods as we mostly require positive numbers here.

File: metrics_ci_integration/lib/jenkins/client/model/jenkins_query_limits.dart [hunk 31]
```diff
@@ -0,0 +1,177 @@
...
+
+  /// Creates an empty range specifier with no edges.
+  ///
+  /// Requests with no range-specifier set will fetch the full data list.
+  const JenkinsQueryLimits.empty() : this._();
+
+  /// Parses a [query] with range-specifier to an instance of [JenkinsQueryLimits].
+  ///
+  /// Throws [FormatException] if parsing fails.
+  factory JenkinsQueryLimits.fromQuery(String query) {
```
Comment (vova-beloded) [inline metrics_ci_integration/lib/jenkins/client/model/jenkins_query_limits.dart]: Should this function throw an ArgumentException if the upper bound is less then lower bound? I mean, if the `{3,1}` query passed, for example. Also, maybe we can add a test for this case?

File: metrics_ci_integration/test/jenkins/client/model/jenkins_build_test.dart [hunk 32]
```diff
@@ -0,0 +1,73 @@
...
+    };
+
+    final jenkinsBuild = JenkinsBuild(
+      number: 1,
+      duration: const Duration(seconds: 10),
+      timestamp: timestamp,
+      result: 'FAILED',
+      url: 'url',
+      artifacts: JenkinsArtifactsResources.artifacts,
+    );
```
Comment (vova-beloded) [inline metrics_ci_integration/test/jenkins/client/model/jenkins_build_test.dart]:
```suggestion
     final timestamp = DateTime(2020);
    const buildNumber = 1;
    const buildDuration = Duration(seconds: 10);
    const buildResult = 'FAILED';
    const buildUrl = 'url';

    final buildJson = {
      'number': buildNumber,
      'duration': buildDuration.inSeconds,
      'timestamp': timestamp.millisecondsSinceEpoch,
      'result': buildResult,
      'url': buildUrl,
      'artifacts': JenkinsArtifactsResources.artifactsJson,
    };

    final jenkinsBuild = JenkinsBuild(
      number: buildNumber,
      duration: buildDuration,
      timestamp: timestamp,
      result: buildResult,
      url: buildUrl,
      artifacts: JenkinsArtifactsResources.artifacts,
    );
```
Can we extract variables from jsons to make the tests more clean and dry ?

File: metrics_ci_integration/test/jenkins/client/util/number_validator_test.dart [hunk 33]
```diff
@@ -0,0 +1,58 @@
...
+    });
+
+    test(".checkValue() should validate the given number", () {
+      expect(() => NumberValidator.checkPositive(1), returnsNormally);
+    });
+
+    test(
+      ".checkRange() should throw ArgumentError if the begin value is null",
+      () {
+        expect(() => NumberValidator.checkPositiveRange(null, 1), throwsArgumentError);
```
Comment (vova-beloded) [inline metrics_ci_integration/test/jenkins/client/util/number_validator_test.dart]: The line length should not be greater than 80.

**Comment by vova-beloded** [comment]
LGTM, @vladlaba !

File: metrics_ci_integration/test/jenkins/resources/jenkins_artifacts_resources.dart [hunk 34]
```diff
@@ -0,0 +1,52 @@
+import 'package:ci_integration/jenkins/client/model/jenkins_build_artifact.dart';
+
+/// A class providing a test data for [JenkinsBuildArtifact].
+class JenkinsArtifactsResources {
```
Comment (illia-romanenko) [inline metrics_ci_integration/test/jenkins/resources/jenkins_artifacts_resources.dart]:
```suggestion
class JenkinsArtifactsTestData {
```
? also what if we put it to `test_data` package?

**Comment by vladlaba** [comment]
@illia-romanenko, @vova-beloded, thanks for your suggestions, guys! 💯
